File: frontend/src/lib/departments.ts
/**
 * Mapping of department codes to full names by college.
 * Based on Boston University's official department abbreviations.
 * 
 * Note: Some codes appear in multiple colleges with different meanings:
 * - AR: Archaeology (CAS) vs Visual Arts (CFA)
 * - CI: Cinema & Media Studies (CAS) vs Cinema & Media Studies (COM)
 * - EC: Economics (CAS) vs Electrical & Computer Engineering (ENG)
 * - ME: Middle East & North Africa Studies (CAS) vs Music Education (CFA) vs Mechanical Engineering (ENG)
 * - MS: Medical Science (CAS) vs Materials Science & Engineering (ENG)
 * - MU: Music (CAS) vs Music (CFA)
 * - BF: Bioinformatics (CDS) vs Bioinformatics (ENG)
 */
export const DEPARTMENT_NAMES: Record<string, Record<string, string>> = {
  CAS: {
    "AA": "African American & Black Diaspora Studies",
    "AH": "History of Art & Architecture",
    "AI": "Asian Studies",
    "AM": "American & New England Studies, including Preservation Studies",
    "AN": "Anthropology",
    "AR": "Archaeology",
    "AS": "Astronomy",
    "BB": "Biochemistry & Molecular Biology",
    "BI": "Biology",
    "C": "Core Curriculum",
    "CG": "Modern Greek",
    "CH": "Chemistry",
    "CI": "Cinema & Media Studies",
    "CL": "Classical Studies",
    "CS": "Computer Science",
    "EC": "Economics",
    "EE": "Earth & Environment",
    "EI": "Editorial Studies",
    "EN": "English",
    "FR": "French Studies (International Programs only)",
    "FY": "First Year Experience",
    "HI": "History",
    "HU": "Humanities",
    "ID": "Interdisciplinary Studies",
    "IN": "Internships",
    "IP": "International Programs",
    "IR": "International Relations",
    "IT": "Italian Studies (International Programs only)",
    "JS": "Jewish Studies",
    "LA": "Hausa",
    "LC": "Chinese",
    "LD": "Amharic, Igbo, Mandinka/Bambara, Setswana/Sesotho, isiZulu, Other African Languages and Linguistics",
    "LE": "Swahili (Kiswahili)",
    "LF": "French",
    "LG": "German",
    "LH": "Hebrew",
    "LI": "Italian",
    "LJ": "Japanese",
    "LK": "Korean",
    "LL": "Language Learning",
    "LM": "isiXosha",
    "LN": "Hindi-Urdu",
    "LO": "Yoruba",
    "LP": "Portuguese",
    "LR": "Russian",
    "LS": "Spanish",
    "LT": "Turkish",
    "LU": "Pulaar",
    "LW": "Wolof, Akan Twi",
    "LX": "Applied Linguistics",
    "LY": "Arabic",
    "LZ": "Persian (Farsi)",
    "MA": "Mathematics & Statistics",
    "ME": "Middle East & North Africa Studies",
    "MR": "Marine Science",
    "MS": "Medical Science",
    "MU": "Music",
    "NE": "Neuroscience",
    "NG": "Nigerian Studies (International Programs only)",
    "NS": "Natural Sciences",
    "PH": "Philosophy",
    "PO": "Political Science",
    "PS": "Psychological & Brain Sciences",
    "PY": "Physics",
    "QU": "Spanish Studies (Quito only)",
    "RN": "Religion",
    "RO": "Romance Studies",
    "SO": "Sociology",
    "SS": "Social Sciences",
    "SY": "Senior-Year Development",
    "TL": "Literary Translation",
    "WR": "Writing",
    "WS": "Women's, Gender & Sexuality Studies",
    "XL": "Comparative Literature",
  },
  QUESTROM: {
    "AC": "Accounting",
    "AD": "Art & Design",
    "BA": "Business Administration",
    "FE": "Finance",
    "IS": "Information Systems",
    "MK": "Marketing",
    "OM": "Operations Management",
    "QM": "Quantitative Methods",
    "SI": "Social Innovation",
    "SM": "Sports Management",
  },
  CDS: {
    "BF": "Bioinformatics",
    "DS": "Data Science",
    "DX": "Online Data Science course",
  },
  CFA: {
    "AR": "Visual Arts",
    "TH": "Theatre",
    "FAC": "FA Courses",
    "ME": "Music Education",
    "ML": "Applied Lessons",
    "MP": "Performance",
    "MT": "Music Theory",
    "MU": "Music",
  },
  CGS: {
    "HU": "Humanities",
    "MA": "Mathematics",
    "NS": "Natural Science",
    "RH": "Rhetoric",
    "SS": "Social Science",
  },
  COM: {
    "CI": "Cinema & Media Studies",
    "CM": "Mass Communication, Advertising & Public Relations",
    "CO": "Communication Core Courses",
    "EM": "Emerging Media Studies",
    "FT": "Film & Television",
    "JO": "Journalism",
  },
  ENG: {
    "BE": "Biomedical Engineering",
    "BF": "Bioinformatics",
    "EC": "Electrical & Computer Engineering",
    "EK": "Engineering Core",
    "ME": "Mechanical Engineering",
    "MS": "Materials Science & Engineering",
    "SE": "Systems Engineering",
  },
  SAR: {
    "HF": "Health Sciences",
    "HS": "Health Sciences",
    "OT": "Occupational Therapy",
    "PT": "Physical Therapy",
    "SH": "Speech & Hearing",
  },
};

/**
 * Get the full name for a department code, optionally scoped by college.
 * 
 * If college is provided, looks up the code in that college's mapping.
 * If college is not provided, searches all colleges and returns the first match
 * (for backwards compatibility with existing code).
 * 
 * @param code - Department code (e.g., "EC", "CS")
 * @param college - Optional college code (e.g., "CAS", "ENG")
 * @returns Full department name, or the code itself if not found
 */
export function getDepartmentName(code: string, college?: string): string {
  if (college && DEPARTMENT_NAMES[college]?.[code]) {
    return DEPARTMENT_NAMES[college][code];
  }
  
  // Fallback: search all colleges (for backwards compatibility)
  // This handles cases where college is not provided
  for (const collegeMap of Object.values(DEPARTMENT_NAMES)) {
    if (collegeMap[code]) {
      return collegeMap[code];
    }
  }
  
  return code;
}

/**
 * Get a display string with both code and name.
 * 
 * @param code - Department code (e.g., "EC", "CS")
 * @param college - Optional college code (e.g., "CAS", "ENG")
 * @returns Display string like "EC - Economics" or "EC - Electrical & Computer Engineering"
 */
export function getDepartmentDisplay(code: string, college?: string): string {
  const name = getDepartmentName(code, college);
  return name !== code ? `${code} - ${name}` : code;
}

File: src/utils/departments.py
"""
Department code to name mapping and search aliases.
Used for resolving queries like "math" -> MA, "computer science" -> CS.
"""

# Flattened code -> name for search text enhancement
# Based on Boston University department abbreviations
DEPT_CODE_TO_NAME: dict[str, str] = {
    "AA": "African American Black Diaspora Studies",
    "AH": "History of Art Architecture",
    "AI": "Asian Studies",
    "AM": "American New England Studies",
    "AN": "Anthropology",
    "AR": "Archaeology",
    "AS": "Astronomy",
    "BB": "Biochemistry Molecular Biology",
    "BI": "Biology",
    "CH": "Chemistry",
    "CI": "Cinema Media Studies",
    "CL": "Classical Studies",
    "CS": "Computer Science",
    "EC": "Economics",
    "EE": "Earth Environment",
    "EN": "English",
    "HI": "History",
    "MA": "Mathematics Statistics",
    "ME": "Middle East North Africa Studies",
    "MU": "Music",
    "NE": "Neuroscience",
    "PH": "Philosophy",
    "PO": "Political Science",
    "PS": "Psychological Brain Sciences",
    "PY": "Physics",
    "RN": "Religion",
    "SO": "Sociology",
    "WR": "Writing",
    "WS": "Women Gender Sexuality Studies",
    "LF": "French",
    "LG": "German",
    "LI": "Italian",
    "LS": "Spanish",
    "LR": "Russian",
    "LY": "Arabic",
    "LC": "Chinese",
    "LJ": "Japanese",
    "LK": "Korean",
    "AC": "Accounting",
    "BA": "Business Administration",
    "BE": "Biomedical Engineering",
    "BF": "Bioinformatics",
    "DS": "Data Science",
    "FT": "Film Television",
    "JO": "Journalism",
}

# Aliases: lowercase search term -> list of department codes
# Include both codes (AH, CS, MA) and full/partial names
DEPT_ALIASES: dict[str, list[str]] = {
    # Common shorthand
    "math": ["MA"],
    "mathematics": ["MA"],
    "stats": ["MA"],
    "statistics": ["MA"],
    "computer": ["CS"],
    "computer science": ["CS"],
    "physics": ["PY"],
    "chemistry": ["CH"],
    "biology": ["BI"],
    "economics": ["EC"],
    "econ": ["EC"],
    "psychology": ["PS"],
    "psych": ["PS"],
    "philosophy": ["PH"],
    "history": ["HI"],
    "english": ["EN"],
    "sociology": ["SO"],
    "anthropology": ["AN"],
    "political": ["PO"],
    "polisci": ["PO"],
    "political science": ["PO"],
    "music": ["MU"],
    "writing": ["WR"],
    "accounting": ["AC"],
    "finance": ["FE"],
}

# Add each code (AH, CS, MA, etc.) and its full name as aliases
for code, name in DEPT_CODE_TO_NAME.items():
    code_key = code.lower()
    if code_key not in DEPT_ALIASES:
        DEPT_ALIASES[code_key] = [code]
    elif code not in DEPT_ALIASES[code_key]:
        DEPT_ALIASES[code_key] = [code] + [c for c in DEPT_ALIASES[code_key] if c != code]
    # Full department name (normalized: lowercase, no & or punctuation)
    name_norm = " ".join(name.replace("&", " ").replace(",", " ").split()).lower()
    if name_norm and name_norm not in DEPT_ALIASES:
        DEPT_ALIASES[name_norm] = [code]
    elif name_norm and code not in DEPT_ALIASES.get(name_norm, []):
        DEPT_ALIASES.setdefault(name_norm, []).insert(0, code)


def get_department_name(code: str) -> str:
    """Return full department name for a code, or the code if unknown."""
    return DEPT_CODE_TO_NAME.get(code.upper(), code)


def resolve_department_query(query: str) -> tuple[list[str], str] | None:
    """
    If the query (or a part of it) is a department alias, return (subject_codes, remaining_query).
    Matches: exact code (AH, CS), exact alias (math, computer science), full name, or name prefix.
    E.g. "math" -> (["MA"], ""), "History of Art" -> (["AH"], ""), "MA" -> (["MA"], "").
    Returns None if no department alias found.
    """
    q = query.strip()
    if not q:
        return None
    q_lower = q.lower()
    # Exact alias or code
    if q_lower in DEPT_ALIASES:
        return (DEPT_ALIASES[q_lower], "")
    if q.upper() in DEPT_CODE_TO_NAME:
        return ([q.upper()], "")
    # Query is prefix of a multi-word department name (e.g. "history of art" matches "history of art architecture")
    # Do NOT match when alias is a short code (ec, cs) - "EC201" and "CS411" are course codes, not dept-only
    best_match: tuple[str, list[str]] | None = None
    for alias_key, codes in DEPT_ALIASES.items():
        if alias_key.startswith(q_lower):
            # Query is prefix of alias - e.g. "history of art" matches "history of art architecture"
            if best_match is None or len(alias_key) > len(best_match[0]):
                best_match = (alias_key, codes)
        elif " " in alias_key and q_lower.startswith(alias_key):
            # Alias is multi-word and query extends it (uncommon)
            if best_match is None or len(alias_key) > len(best_match[0]):
                best_match = (alias_key, codes)
    if best_match:
        return (best_match[1], "")
    # Check first word/token
    parts = q.split()
    if parts:
        first = parts[0].lower()
        if first in DEPT_ALIASES:
            remaining = " ".join(parts[1:]).strip()
            return (DEPT_ALIASES[first], remaining)
        if parts[0].upper() in DEPT_CODE_TO_NAME:
            remaining = " ".join(parts[1:]).strip()
            return ([parts[0].upper()], remaining)
    return None
